Add counterfactual minus guided/unguided diff at location level

Zapiano opened the pull request:
Followed an approach similar to https://github.com/open-AIMS/ADRIA.jl/issues/452#issuecomment-1942888428 but for a spatial level.

Note: Depends on #847 

## Example of code

```julia
using ADRIA
using WGLMakie, GeoMakie, GraphMakie

Makie.inline!(false)

# Load domain
dom = ADRIA.load_domain(...)

# Create scenarios
num_scens = 2^10
scens = ADRIA.sample(dom, num_scens)

# Run model / Load ResultSet
rs = ADRIA.run_scenarios(dom, scens, "45")

# Extract cf_difference_loc for guided and unguided (in that order)
rc = ADRIA.metrics.relative_cover(rs)

# Compute difference between guided and counterfactual using the 0.6-th quantile
gd_res = metrics.ensemble_loc_difference(rc, scens; agg_metric=0.6)

# Compute difference between unguided and counterfactual using the median
ug_res = metrics.ensemble_loc_difference(r_cover, scens; diff_target=:unguided)

# Plot maps of difference to the counterfactual
ADRIA.viz.map(rs, gd_res[summary=At(:agg_value)]; diverging=true)
ADRIA.viz.map(rs, ug_res[summary=At(:agg_value)]; diverging=true)
```

## Plots
- 1024 scenarios

| Guided - Counterfactual (0.6-th quantile) | Unguided - Counterfactual (median)|
| ----------------------- | ------------------------- |
| ![gd_cf_diff_map](https://github.com/user-attachments/assets/200f035c-4183-44e9-a590-2fae7fe1f30d) | ![ug_cf_diff_map](https://github.com/user-attachments/assets/7728e1ff-a8fb-482b-b800-ff822fb51e4a) |

ConnectedSystems commented on src/metrics/spatial.jl:
```diff
@@ -1,3 +1,6 @@
+using Bootstrap
+using Random
+
 """Functions and methods to produce location-level summaries."""
```
High-level docstrings should be at the very top of file.

ConnectedSystems commented on src/metrics/spatial.jl:
```diff
@@ -106,3 +109,40 @@ function summarize(
+        ug_diff = collect(ug_outcomes[loc, ug_shuf_set] .- cf_outcomes[loc, cf_shuf_set1])
+        ug_result[loc] = bootstrap(median, ug_diff, BalancedSampling(100)).t0[1]
+
+        cf_shuf_set2 = shuffle(1:n_scens)
+        gd_shuf_set = shuffle(1:n_scens)
+        gd_diff = collect(gd_outcomes[loc, gd_shuf_set] .- cf_outcomes[loc, cf_shuf_set2])
```
Could reuse the first `cf_shuf` here

ConnectedSystems commented on src/metrics/spatial.jl:
```diff
@@ -106,3 +109,40 @@ function summarize(
+"""
+    cf_difference_map(rs, scens, metric)
+
+Give the mean bootstraped difference from the counterfactual for a given metric.
```
```suggestion
Give the mean bootstrapped difference from the counterfactual for a given metric.
```

ConnectedSystems replied:
Document what is returned?

ConnectedSystems commented on src/metrics/spatial.jl:
```diff
@@ -106,3 +109,40 @@ function summarize(
+
+    n_locs = length(outcomes.sites)
+
+    ug_result = ZeroDataCube(; T=Float64, sites=locations.reef_siteid)
+    gd_result = ZeroDataCube(; T=Float64, sites=locations.reef_siteid)
+    n_scens = size(cf_outcomes, :scenarios)
```
Not interested in the confidence bounds?

ConnectedSystems commented:
Bump @Zapiano

ConnectedSystems commented:
How close are we to merging this?

Zapiano commented:
> How close are we to merging this?

I'll start looking into this now and let you know what's missing or of it's ready soon.

Zapiano commented:
@ConnectedSystems This one is ready for another review.

ConnectedSystems commented:
An example usage would be appreciated too @Zapiano

Zapiano commented:
> Thanks.
> 
> An example usage would be appreciated too @Zapiano

I've added an example in this PR. Once this PR is approved I can add this plot to the documentation on a separate PR.

ConnectedSystems commented:
I'm not sure the indicated use it the best approach.

It forces only one set of bootstrapped results (only the median?), and forces additional computations to be done (e.g., what if I only want `gd_res`?)

```julia
# Extract cf_difference_loc for guided and unguided (in that order)
rc = ADRIA.metrics.relative_cover(rs)
gd_res, ug_res = ADRIA.metrics.cf_difference_loc(rc, scens)  # <- this method name feels awkward

# Plot cf_difference_loc for unguided and guided
gd_fig = ADRIA.viz.diff_map(rs, gd_res[2, :])  # <- why `[2, :]`? (I know why, but feels awkward)
ug_fig = ADRIA.viz.diff_map(rs, ug_res[2, :])
```

Instead, I suggest something like the below which, to me, is a cleaner UX.

```julia
rc = ADRIA.metrics.relative_cover(rs)

# outcome, scenario spec, percentile value of interest
gd_res = ADRIA.metrics.ensemble_difference(rc, scens, 0.5; timestep=2, scenario_type=:guided)

# outcome, scenario spec, statistical function of interest
ug_res = ADRIA.metrics.ensemble_difference(rc, scens, mean; timestep=2, scenario_type=:unguided)
```

On the plotting, it kind of feels like we're getting to the point where we want our own YAXArray subtype. 

```julia
ADRIA.viz.map(rs, gd_res::SomeCustomYAXArray?)  #<- uses different defaults to the "usual" `viz.map()`
```

Zapiano commented:
@ConnectedSystems this was updated to address your comments. The description was updated with new example of usage and plots.

Zapiano commented:
> On the plotting, it kind of feels like we're getting to the point where we want our own YAXArray subtype.
> 
> ```julia
> ADRIA.viz.map(rs, gd_res::SomeCustomYAXArray?)  #<- uses different defaults to the "usual" `viz.map()`
> ```

I'm not sure I understand the purpose of having our own YAXArray subtype.

ConnectedSystems commented:
> I'm not sure I understand the purpose of having our own YAXArray subtype.

To be able to define our own custom styling for specific use cases and be able to make different styles when plotting some arbitrary results (e.g., `some_plot(a_matrix)`) compared to `some_plot(a_general_yaxarray)` and `some_plot(a_yaxarray_with_context_specific_metadata)`

ConnectedSystems commented:
@Zapiano fyi there are conflicts now because of the merge of #863. Could you rebase or merge main as appropriate?

Zapiano commented:
> @Zapiano fyi there are conflicts now because of the merge of #863. Could you rebase or merge main as appropriate?

I rebased but the tests are not passing. I'll let you know when this is ready again.

Zapiano commented:
> > @Zapiano fyi there are conflicts now because of the merge of #863. Could you rebase or merge main as appropriate?
> 
> I rebased but the tests are not passing. I'll let you know when this is ready again.

@ConnectedSystems  the tests that were not passing is a problem on the main branch apparently (https://github.com/open-AIMS/ADRIA.jl/issues/890). If you prefer waiting for that to be fixed so we are sure the tests are passing for this PR as well it's fine, otherwise this one is ready to merge.

ConnectedSystems commented:
Have you updated to the latest dependencies?

Zapiano commented:
> Have you updated to the latest dependencies?

Yes.

ConnectedSystems commented:
> I'm not sure I understand the purpose of having our own YAXArray subtype.

FYI, Makie's theme recipes seem more stable now, and better documented:

https://docs.makie.org/stable/tutorials/wrap-existing-recipe

ConnectedSystems commented:
This PR can be merged once #890 is confirmed to be addressed by #891

Zapiano commented:
@ConnectedSystems tests are passing and I've updated testset names for the tests that are related to this PR as we discussed.

Zapiano commented:
@ConnectedSystems just reminding you that this is ready for review.

ConnectedSystems commented on src/metrics/spatial.jl:
```diff
@@ -122,3 +125,105 @@ function summarize(
+gd_res = metrics.ensemble_loc_difference(r_cover, scens; agg_metric=0.6)
+
+# Compute difference between unguided and counterfactual using the median
+ug_res = metrics.cf_difference_loc(r_cover, scens; diff_target=:unguided)
```
Incorrect example here?

Zapiano replied:
Fixed.

ConnectedSystems commented on src/metrics/spatial.jl:
```diff
@@ -122,3 +125,105 @@ function summarize(
+# Compute difference between unguided and counterfactual using the median
+ug_res = metrics.cf_difference_loc(r_cover, scens; diff_target=:unguided)
+
+# Plot maps of difference to the counterfactual
```
```suggestion
# Plot maps of difference to the counterfactual at timestep 2
```

Is the above correct?

Zapiano replied:
Yes, this `2` is not timestep, `gd_res[1,:]` and `gd_res[3,:]` are the lower and upper bounds of the confidence interval and `gd_res[2,:]` is the metric used (median or whatever).

ConnectedSystems replied:
Hmm, is `gd_res` a YAXArray? It would be clearer what the "2" is for the example if the axis name is used as the selector instead

ConnectedSystems replied:
I'd ask that the comment be updated to clarify what the "2" is.

ConnectedSystems commented on src/metrics/spatial.jl:
```diff
@@ -122,3 +125,105 @@ function summarize(
+    agg_metric::Union{Function,AbstractFloat}=median,
+    diff_target=:guided,
+    conf::Float64=0.95,
+    rand_seed=1234
```
Minor point - could we use a global random seed instead?

```suggestion
    rng::AbstractRNG=Random.GLOBAL_RNG
```

Later on...

`shuffle(rng, 1:n_cf_outcomes)[1:min_n_outcomes]`

Zapiano replied:
Does that mean that the user shouldn't be allowed to set the seed and I should also remove this `Random.seed!(rand_seed)` ?

ConnectedSystems replied:
Yes, you should remove the `Random.seed!` inside the function.

It means the user can set it outside the function.

This way the state of the random number generator does not "leak" outside the scope of the function.
Currently, the function "resets" the seed state, which may affect other functions.

Best practice is to set a RNG and pass that to each function if necessary

```julia
rng = Random.Xoshiro(123)  # use a different RNG algorithm
some_function(args...; rng=rng)
another_function(args...; rng=rng)
```

As an aside: the current exception in ADRIA is the scenario running, where I made the decision that we want the outcome to be tied to the scenario factors. But it's implemented in a suboptimal way... another thing to revisit.

Zapiano replied:
Makes sense, thanks for the explanation.

ConnectedSystems reviewed (changes requested):
Two minor comments, but this is ready to be merged once those are addressed.

Zapiano commented:
> Two minor comments, but this is ready to be merged once those are addressed.

Done, and all tests passing.

Zapiano commented:
@ConnectedSystems I made two changes:

1. I updated the example (and tests) to use the YAXArray dim/label names instead of the index `2` directly when assessing the aggregate metric 
2. I have deleted the function `diff_map` and I'm using `map` instead, with an option for `diverging` color palette. That seems more flexible if we want to plot other maps that also need this palette in the future but are not "diff maps".
